doc/tally-recovery.md

```diff
@@ -0,0 +1,95 @@
+# Recovery of the corrected vote tally from fragment logs
+
+In the Fund 4 time frame, we've been unable to completely resolve issues
+with the Catalyst app usage of spending counters in witnesses of successive
+ballot (aka VoteCast) transactions for the single voting account.
+This may lead to ballot transactions legitimately submitted by voting users
+to be rejected by the consensus leader nodes at block creation.
+
+To allow computation of the tally across all ballots legitimately submitted
+by the users in face of such discrepancies, an alternative
+tallying method was devised and implemented in the catalyst-toolbox tool.
+Afterwards, this method can be used to confirm the blockchain tally did not
+miss any received legitimate ballots that were recorded into persistent
+fragment logs.
+
+## Input: persistent fragment logs
+
+The node or nodes serving REST requests from the clients submitting ballots
+has collection of persistent fragment logs enabled. These log files record all
+blockchain fragments received by the node and admitted to its mempool,
+meaning that the fragments are deduplicated, correctly formatted, and pass some
+minimal sanity checks, but otherwise their validity for inclusion into a block
+is not verified. The timestamp of fragment arrival is stored along with the
+fragment; in the alternative tallying process described here, the chronological
+order of fragments is used to decide which choice applies when multiple ballots
+for the same proposal have been received from a single account.
+
+The format of the log entries is bincode serialization of a Rust structure
+containing the unsigned integer timestamp in seconds since the Unix epoch,
+and the fragment serialized in the binary blockchain format: 
+
+```rust
+pub struct FragmentLogEntry {
+    /// The time this fragment was registered and accepted by the pool
+    pub time: SecondsSinceUnixEpoch,
+    /// Fragment body serialized as byte array
+    pub fragment: RawFragment,
+}
+```
+
+## Initial state: voter accounts in block0
+
+To determine which accounts can legitimately vote, the tool parses the
+genesis block of the Catalyst blockchain. It is assumed that the registered
+voters' account balances are created in the initial fund distribution in the
+genesis block. The initial ledger state is recreated from the genesis block
+just like in the blockchain node (with some technical twists: a
+mirror set of accounts is created because we need some private keys to
+supply to the library code performing ledger state transitions).
+
+## Replay of transactions
+
+The tally recovery command parses the persistent fragment logs and applies
+the fragments in order they were received, to replay changes in the ledger state
+of a blockchain node that would process transactions to the same effect, if
+the account spending counters used to sign transactions spending form the same
```

Comment:
This last sentence is a bit confusing to me.

Reply:
Tried to fix.